activation/activation.go

```diff
@@ -627,15 +630,15 @@ func (b *Builder) createAtx(ctx context.Context) (*types.ActivationTx, error) {
 
 	// the following method waits for a PoET proof, which should take ~1 epoch
 	expireLayer := (pubEpoch + 2).FirstLayer()
-	atxExpired := b.layerClock.AwaitLayer(expireLayer) // this fires when the target epoch is over
+	atxExpiredContext := b.layerClock.AwaitLayer(ctx, expireLayer) // this fires when the target epoch is over
 	commitmentAtx, err := b.getCommitmentAtx(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("getting commitment atx failed: %w", err)
 	}
 
 	b.log.With().Info("building NIPost", log.Stringer("pub_epoch", pubEpoch), log.Stringer("expire_layer", expireLayer))
 
-	nipost, err := b.nipostBuilder.BuildNIPost(ctx, hash, *commitmentAtx, atxExpired)
```

Comment:
this ctx is the app's top level context. it is used throughout the app goroutines to listen for app shutdown.
should not combine this with whether an ATX should expire.

Reply:
in this specific case, the code return immediately in both cases. but this is not a general case. what is so wrong with waiting for two signals (layer arrival / shutdown) separately?

Reply:
>this ctx is the app's top level context. it is used throughout the app goroutines to listen for app shutdown.

`AwaitLayer` creates a new context with an additional constraint to indicate when the layer has been reached. This new context can now be used in `BuildNIPost` to signal when it is supposed to stop creating NIPosts. The original `ctx` without this constraint is still available and can continued to be used. Contexts eventually form a tree:

![image](https://user-images.githubusercontent.com/5011972/195433955-50ae5081-dda4-47da-965c-005e37afe1de.png)

Cancelling a parent context cancels all child contexts but not vice versa! The root context is untouched and is continued to be used by other routines that only listen for app shutdown.

Reply:
> in this specific case, the code return immediately in both cases. but this is not a general case. what is so wrong with waiting for two signals (layer arrival / shutdown) separately?

Contexts are all about specific cases. It is not wrong for a function to create a new context from the context it receives for every sub-function it calls:

```golang
func LongRunningFunc(ctx context.Context) {
	child1Ctx, cancelFunc1 := context.WithCancel(ctx)
        go someFunctionThatNeedsToBeCancelled(child1Ctx)

	// some logic that possibly calls cancelFunc1
        child2Ctx, cancelFunc2 := context.WithTimeout(rootCtx, 1 * time.Second)
	defer cancelFunc2()
	someFunctionThatMustReturnQuickly(child2Ctx)
}
```

This way if `rootCtx` gets canceled `someFunctionThatMustReturnQuickly` and `someFunctionThatNeedsToBeCancelled` receive this signal too without the need to pass in 2 contexts

Reply:
thanks for the illustration :P

i agree with you that context is all about specific case. but i also think it makes the code more complex than it needs to be at the moment.

my rationale:
- major: AwaitLayer is used in many places. in some the new code now insist to finish waiting for layers before the app can exit. this is the main reason i block this PR. for mainnet, the layer duration is set to be 5 min. we don't want the app shutdown to hang for that long.

- moderate: i think this should be dealt with in a case-by-case manner in each component. the original interface uses channel, which is very simple. and it's up to the component to determine its action. 
take this for example https://github.com/spacemeshos/go-spacemesh/blob/develop/activation/activation.go#L648-L652
```
select {
	case <-ctx.Done():
		return nil, fmt.Errorf("failed to wait for publication epoch: %w", err)
	case <-b.layerClock.AwaitLayer(pubEpoch.FirstLayer()):
}
```
can be rewritten to 
```
select {
	case <-ctx.Done():
		return nil, fmt.Errorf("failed to wait for publication epoch: %w", err)
	case <-b.layerClock.AwaitLayer(pubEpoch.FirstLayer()):
	    if ctx.Error() != nil {
	        return nil, fmt.Errorf("context error: %w", ctx.Error)
	    }
}
```
to differentiate the two. 

in the case where you handle them almost the same
```
	select {
	case poetProofRef = <-nb.poetDB.SubscribeToProofRef(nb.state.PoetServiceID, nb.state.PoetRound.ID):
	case <-atxExpired:
		nb.poetDB.UnsubscribeFromProofRef(nb.state.PoetServiceID, nb.state.PoetRound.ID)
		return nil, fmt.Errorf("%w: while waiting for poet proof, target epoch ended", ErrATXChallengeExpired)
	case <-ctx.Done():
		return nil, ErrStopRequested
	}
```
i simply cannot see why this code is offensive. if all 3 channels unblocks at the same time, the first case may add an extra check for ctx.Error(). but even if it doesn't, it will stop at some later check.

- minor: for a reader trying to understand the code, tracing exactly where context is updated is tedious. i always feel using a separate context for p2p request timeout is a cleaner separation for app logic error handling.

Reply:
> AwaitLayer is used in many places. in some the new code now insist to finish waiting for layers before the app can exit.

This is not the case. `AwaitLayer` returns when the parent context is canceled. The context signals to `AwaitLayer` that it should stop waiting. These two are functionally identical:

`AwaitLayer` without context:
```golang
select {
	case <-ctx.Done():
		return nil, fmt.Errorf("failed to wait for publication epoch: %w", err)
	case <-b.layerClock.AwaitLayer(pubEpoch.FirstLayer()):
	    if ctx.Error() != nil {
	        return nil, fmt.Errorf("context error: %w", ctx.Error)
	    }
}
```

`AwaitLayer` with context:
```golang
<-b.layerClock.AwaitLayer(ctx, pubEpoch.FirstLayer())
select {
	case <-ctx.Done():
		return nil, fmt.Errorf("failed to wait for publication epoch: %w", err)
	default:
}
```

Reply:
> This is not the case. AwaitLayer returns when the parent context is canceled.

i stand corrected. you are right.
but i still think this code is not better than 

```
select {
	case <-ctx.Done():
		return nil, fmt.Errorf("failed to wait for publication epoch: %w", err)
	case <-b.layerClock.AwaitLayer(pubEpoch.FirstLayer()):
	    if ctx.Error() != nil {
	        return nil, fmt.Errorf("context error: %w", ctx.Error)
	    }
}
```
and in this case, you can keep `AwaitLayer()` simple with channel, and not using context that dev can potentially misuse.